drm-howto/modeset-atomic.c

```diff
@@ -0,0 +1,1114 @@
+	ev.page_flip_handler = modeset_page_flip_event;
+
+	while (output_list) {
+		/* remove from global list */
+		iter = output_list;
+		output_list = iter->next;
+
+		/* if a page-flip is pending, wait for it to complete */
+		iter->cleanup = true;
+		fprintf(stderr, "wait for pending page-flip to complete...\n");
+		while (iter->pflip_pending) {
+			ret = drmHandleEvent(fd, &ev);
+			if (ret)
+				break;
+		}
+
+		/* restore saved CRTC configuration */
+		if (!iter->pflip_pending)
+			drmModeSetCrtc(fd,
+				       iter->saved_crtc->crtc_id,
+				       iter->saved_crtc->buffer_id,
+				       iter->saved_crtc->x,
+				       iter->saved_crtc->y,
+				       &iter->conn_id,
+				       1,
+				       &iter->saved_crtc->mode);
+		drmModeFreeCrtc(iter->saved_crtc);
+
+		/* destroy framebuffers */
+		modeset_destroy_fb(fd, &iter->bufs[1]);
+		modeset_destroy_fb(fd, &iter->bufs[0]);
+
+		/* free allocated memory */
+		free(iter);
+	}
+}
+
+/*
+ * main() also changes. Instead of performing the KMS setup calling
+ * drmModeSetCrtc(), we instead setup it using the atomic API with
+ * the function modeset_perform_modeset().
+ */
+
+int main(int argc, char **argv)
+{
+	int ret, fd;
+	const char *card;
+	struct modeset_out *iter;
+	drmModeAtomicReq *req;
+
+	/* check which DRM device to open */
+	if (argc > 1)
+		card = argv[1];
+	else
+		card = "/out/dri/card0";
+
+	fprintf(stderr, "using card '%s'\n", card);
+
+	/* open the DRM device */
+	ret = modeset_open(&fd, card);
+	if (ret)
+		goto out_return;
+
+	/* prepare all connectors and CRTCs */
+	ret = modeset_prepare(fd);
+	if (ret)
+		goto out_close;
+
+	/* save crtc object, so we can restore it later */
+	for (iter = output_list; iter; iter = iter->next)
+		iter->saved_crtc = drmModeGetCrtc(fd, iter->crtc_id);
+
+	/* perform modeset using atomic commit */
+	req = drmModeAtomicAlloc();
+	modeset_perform_modeset(fd, req);
+	drmModeAtomicFree(req);
+
+	/* draw some colors for 5seconds */
+	modeset_draw(fd);
+
+	/* cleanup everything */
+	modeset_cleanup(fd);
+
+	ret = 0;
+
+out_close:
+	close(fd);
+out_return:
+	if (ret) {
+		errno = -ret;
+		fprintf(stderr, "modeset failed with error %d: %m\n", errno);
+	} else {
+		fprintf(stderr, "exiting\n");
+	}
+	return ret;
+}
+
+/*
+ * This is a very simple example to show how to use the KMS atomic API and how
+ * different it is from legacy KMS. Most modern drivers are using the atomic
+ * API, so it is important to have this example.
+ *
+ * Just like vsync'ed double-buffering, the atomic API does not not solve all
+ * the problems that can happen  and you have to figure out the best
+ * implementation for your use case.
+ *
+ * If you want to take a look at more complex examples that makes use of KMS
+ * atomic API, I can recommend you:
+ *
+ * - kms-quads: a more complex (but also well-explained) example that can be
+ *   used to learn how to build a compositor using the atomic API. Supports both
+ *   GL and software rendering.
```

Comment:
Maybe links to the repositories would be good here.

Reply:
Done.